### DISCUSSION TITLE
SP2000 Storage Unit

### DISCUSSION THREAD

**Discussion by tonyemon** [johanmeijer/grott#307] [Q&A] [answered]
Hello

I have enjoyed getting the Grott project installed on a raspberry pi and am getting some results from the SP2000 and shine datalogger. The correct Seriel numbers are being recognized and data is being forwarded to the growatt server but the default settings are not pulling out any battery information.

Is it possible to get this working, so that the battery records are pulled in as well. The SP2000 is installed between the PV Array and my Inverter to manage the DC electricity flows between Inverter and battery.

I have tried changing the Inverter Type to SPA but it then fails to recognise the sereil number and list no data on the Journalctl output, but the records are still being forwarded to the Growatt servers.

Please could someone advise if anything can be done.

Thanks

**Comment by tonyemon** [comment]
Hi

I have been looking through the threads and managed to guess a record layout that works for the SP2000 storage unit as per listing below. I have coded this into my local grottconf.py file and it pulls out the data I need. 

 self.recorddict3 = {"T06NNNN": {
            "decrypt"           : {"value" :"True"},
            "datalogserial"     : {"value" :16, "length" : 10, "type" : "text","incl" : "yes"},
            "pvserial"          : {"value" :76, "length" : 10, "type" : "text"},
            "date"              : {"value" :136, "divide" : 10},
            "recortype1"        : {"value" :150, "length" : 2, "type" : "num","incl" : "no"},
            "recortype2"        : {"value" :154, "length" : 2, "type" : "num","incl" : "no"},
            "Generation_status" : {"value" :158, "length" : 2, "type" : "num"},
            "pvpowerin"         : {"value" :162, "length" : 4, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Solar_voltage"     : {"value" :170, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Solar_current"     : {"value" :174, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Solar Generation"  : {"value" :178, "length" : 4, "type" : "num", "divide" : 10,"incl" : "yes"},
            "bat_Volt"          : {"value" :206, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Battery_Charge"    : {"value" :218, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Grid_AC_Voltage"   : {"value" :226, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Import_From_Grid"  : {"value" :238, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Temperature"       : {"value" :242, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "export_to_Grid"    : {"value" :250, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "test_Bat_dischg"   : {"value" :346, "length" : 2, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Solar_today"       : {"value" :358, "length" : 4, "type" : "num", "divide" : 10,"incl" : "yes"},
            "Solar_total"       : {"value" :366, "length" : 4, "type" : "num", "divide" : 10,"incl" : "yes"}

Many thanks for this great program and my call/question can now be closed.

Regards

Tony

**Comment by giangy11** [comment, marked as answer]
@johanmeijer  is it possible to add this configuration to the master version? It could be useful for everyone having an SP2000 storage converter model
thanks a lot

**Reply by tonyemon** [reply]
Hello Johan....please feel free to update your master version. Regards

**Comment by johanmeijer** [comment]
Thanks. I will add this. Do you have a original data record for me so I can test it?

**Reply by giangy11** [reply]
Hi,
Just copied from the grott log, should be this one 

		  <socket.socket fd=32, family=2, type=1, proto=0, laddr=('172.30.33.2', 56804), raddr=('47.91.67.66', 5279)>
	 - Growatt original Data:
		 \x00\x22\x00\x06\x01\x5f\x01\x04\x0d\x22\x2c\x41\x20\x46\x4d\x77\x23\x57\x77
		 \x61\x74\x74\x47\x72\x6f\x77\x61\x74\x74\x47\x72\x6f\x77\x61\x74\x74\x47\x72
		 \x5a\x33\x3b\x4c\x42\x75\x4b\x5f\x40\x24\x74\x74\x47\x72\x6f\x77\x61\x74\x74
		 \x47\x72\x6f\x77\x61\x74\x74\x47\x72\x6f\x77\x76\x7f\x6f\x52\x40\x59\x74\x61
		 \x74\x74\x6b\x72\x6f\x77\x61\x74\x74\x47\x6e\x6f\x77\x61\x74\x74\x47\x72\x6e
		 \x77\x61\x74\x74\x47\x72\x6f\x75\x60\x9d\x74\x47\x72\x6f\x77\x61\x74\x5d\x4e
		 \x7c\x6f\x38\x61\x74\x3c\x02\x72\x6f\x77\x61\x74\x74\x47\xea\x6f\xd7\x61\x1a
		 \x74\x47\x72\x6f\x77\x61\x74\x74\x47\x72\x6f\x77\x61\x74\x74\x4a\x72\x6f\x77
		 \x61\x74\x74\x47\x72\x6f\x77\x61\x74\x74\x47\x72\x6f\x77\x61\x74\x74\x47\x5f
		 \x6f\x2e\x61\x74\x74\x47\x72\x6f\x77\x61\x74\x64\x47\x72\x6c\x4c\x61\x74\x74
		 \x47\x72\x6f\x77\x61\x74\x74\x47\x72\x6f\x77\x61\x74\x74\x47\x72\x6b\x77\x61
		 \x75\xf6\x47\x72\x6f\x72\x61\x74\x75\x93\x72\x6f\x77\x61\x74\x74\x47\x72\x6f
		 \x77\x61\x74\x74\x47\x72\x6f\x77\x64\x74\x72\x47\x72\x6f\x77\x61\x74\x74\x47
		 \x72\x6f\x77\x61\x74\x74\x47\x72\x6f\x77\x61\x74\x74\x47\x72\x6f\x77\x3b\x74
		 \xf2\x47\
	 - Grott automatic protocol detection
	 - Grott data record length 359
	 - layout   :  T060104
	 - no matching record layout found, try generic
	 - Record layout used :  T06NNNN
	 - Growatt data decrypted V2
	 - Grott Growatt data decrypted
	 - Growatt plain data:
		 00220006015f01044a50433641323930513800000000000000000000000000000000000000003
		 5445a383632393037450000000000000000000000000000000000000000170b1b153236030000
		 002c000000000000001c0000000000000001000000000000000201e90000000000000029090e0
		 04f00004845000000000000009800a0006e0000000000000000000000000000000d0000000000
		 000000000000000000000000000000002d0059000000000000000000100000033b00000000000
		 000000000000000000000000000040000018200000005000001d4000000000000000000000000
		 0000000000050006000000000000000000000000000000000000000000000000005a008600000
		 00000000000000000000000000000000000000000000000000000000000000000000000000000
		 00000000000000000000000000000000000000000000000000000000000000000000000000000
		 000000000000000000000e29c
	 - Possible Inverter serial 5DZ862xxxx
	 - Inverter serial not recognised - using inverter type default
	 - Growatt new layout processing
		 - decrypt       :  True
		 - offset        :  6
		 - record layout :  T06NNNN

	 - Grott data record date/time processing started
	 - date-time:  2023-11-27T21:50:54
	 - Grott values retrieved:
		 -  datalogserial        :  JPC6A290Q8
		 -  pvserial             :  5DZ862xxxx
		 -  pvstatus             :  0
		 -  pvpowerin            :  0.0
		 -  pv1voltage           :  2.8
		 -  pv1current           :  0.0
		 -  pv1watt              :  0.0
		 -  pv2voltage           :  0.1
		 -  pv2current           :  0.0
		 -  pv2watt              :  0.0
		 -  pvpowerout           :  13156.1
		 -  pvfrequentie         :  0.0
		 -  pvgridvoltage        :  0.0
		 -  pvgridcurrent        :  0.0
		 -  pvgridpower          :  268929.4
		 -  pvgridvoltage2       :  7.9
		 -  pvgridcurrent2       :  0.0
		 -  pvgridpower2         :  121248153.6
		 -  pvgridvoltage3       :  0.0
		 -  pvgridcurrent3       :  0.0
		 -  pvgridpower3         :  996163.2
		 -  pvenergytoday        :  720896.0
		 -  pvenergytotal        :  0.0
		 -  totworktime          :  0.0
		 -  pvtemperature        :  0.0
		 -  pvipmtemperature     :  0.0
		 -  epv1today            :  1.6
		 -  epv1total            :  82.7
		 -  epv2today            :  0.0
		 -  epv2total            :  0.0
		 -  epvtotal             :  0.0
	 - MQTT jsonmsg: 
			 {"device": "5DZ862xxxx", "time": "2023-11-27T21:50:54", "buffered": "no",
			 "values": {"datalogserial": "JPC6A2zxxx", "pvserial": "5DZ862xxxx",
			 "pvstatus": 0, "pvpowerin": 0, "pv1voltage": 28, "pv1current": 0, "pv1watt":
			 0, "pv2voltage": 1, "pv2current": 0, "pv2watt": 0, "pvpowerout": 131561,
			 "pvfrequentie": 0, "pvgridvoltage": 0, "pvgridcurrent": 0, "pvgridpower":
			 2689294, "pvgridvoltage2": 79, "pvgridcurrent2": 0, "pvgridpower2":
			 1212481536, "pvgridvoltage3": 0, "pvgridcurrent3": 0, "pvgridpower3":
			 9961632, "pvenergytoday": 7208960, "pvenergytotal": 0, "totworktime": 0,
			 "pvtemperature": 0, "pvipmtemperature": 0, "epv1today": 16, "epv1total":
			 827, "epv2today": 0, "epv2total": 0, "epvtotal": 0}}
	 - No MQTT message sent, MQTT disabled
	 - Grott Send data to PVOutput disabled 
	 - Grott Send data to Influx disabled 
	 - Grott extension processing started :  grottext.ha
	 - Grott extension processing ended :  0

	 - Growatt packet received:
		  <socket.socket fd=33, family=2, type=1, proto=0, laddr=('172.30.33.2', 5279), raddr=('192.168.31.197', 11146)>
	 - Data less then minimum record length, data not processed

**Reply by giangy11** [reply]
I tried also adding custom JSON in the config folder but without success
